### PR TITLE
Add cpu profiling start/stop endpoints

### PR DESCRIPTION
**Pull request by nickrobinson251**

### CODE REVIEW

**Comment by nickrobinson251** [comment]
hmmm... at least on localhost this captures mostly _other things_

Screenshot 2022-12-07 at 14 31 25
Screenshot 2022-12-07 at 14 32 13

**Comment by nickrobinson251** [comment]
should just look something like

Screenshot 2022-12-07 at 14 43 24

we're profiling something like `for _ in 1:100 peakflops() end`

File: src/PerformanceProfilingHttpEndpoints.jl [hunk 1]
```diff
@@ -82,45 +74,94 @@ function cpu_profile_endpoint(req::HTTP.Request)
...
+    n = convert(Int, parse(Float64, get(qp, "n", default_n())))
+    delay = parse(Float64, get(qp, "delay", default_delay()))
+    return _start_cpu_profile(n, delay)
+end
+
+function cpu_profile_stop_endpoint(req::HTTP.Request)
+    uri = HTTP.URI(req.target)
+    qp = HTTP.queryparams(uri)
+    with_pprof = parse(Bool, get(qp, "pprof", default_pprof()))
+    return _stop_cpu_profile(with_pprof)
```
Comment (NHDaly) [inline src/PerformanceProfilingHttpEndpoints.jl]: +1 to @quinnj's suggestion to stop the profile as soon as possible, and then do the rest of this work after it's stopped.

**Comment by nickrobinson251** [comment]
okay, i re-ran the profiling via the different endpoints (`/profile` versus `/profile_start`+`/profile_stop`) and see almost-identical profiles, so pretty sure my worry above was just due to poor execution of my experiments

File: src/PerformanceProfilingHttpEndpoints.jl [hunk 2]
```diff
@@ -170,9 +212,10 @@ end
 
 function _start_alloc_profile(sample_rate)
     @info "Starting allocation Profiling from PerformanceProfilingHttpEndpoints with configuration:" sample_rate
+    resp = HTTP.Response(200, "Allocation profiling started.")
     Profile.Allocs.clear()
     Profile.Allocs.start(; sample_rate)
-    return HTTP.Response(200, "Allocation profiling started.")
+    return resp
```
Comment (NHDaly) [inline src/PerformanceProfilingHttpEndpoints.jl:213]: genius

**Review by NHDaly — APPROVED** [review]
LGTM! :) Thanks @nickrobinson251

File: test/runtests.jl [hunk 3]
```diff
@@ -18,26 +18,61 @@ const url = "http://127.0.0.1:$port"
...
 
-        @info "Finished tests, waiting for peakflops workload to finish."
-        done[] = true
-        wait(t)  # handle errors
+            @info "Finished `profile_start/stop` tests, waiting for peakflops workload to finish."
+            done[] = true
+            wait(t)  # handle errors
+
+            # We retrive data via PProf directly if `pprof=true`; make sure that path's tested.
+            # This second call to `profile_stop` shoudl still return the profile, even though
```
Comment (nickrobinson251) [inline test/runtests.jl]:
```suggestion
            # This second call to `profile_stop` should still return the profile, even though
```

**Comment by NHDaly** [comment]
Awesome! 😊

**Pull request merged**